mmdet3d/core/bbox/coders/smoke_bbox_coder.py

```diff
@@ -0,0 +1,172 @@
+import numpy as np
+import torch
+
+from mmdet.core.bbox import BaseBBoxCoder
+from mmdet.core.bbox.builder import BBOX_CODERS
+
+
+@BBOX_CODERS.register_module()
+class SMOKECoder(BaseBBoxCoder):
+    """Bbox Coder for SMOKE.
+
+    Args:
+        base_depth (tuple[float]): Depth references for decode box depth.
+        base_dim (tuple[tuple[float]]): Dimension references for decode
+            box dimension for each category.
+        code_size (int): The dimension of boxes to be encoded. Default: 9.
+    """
+
+    def __init__(self, base_depth, base_dim, code_size):
+        super(BaseBBoxCoder, self).__init__()
+        self.base_depth = base_depth
+        self.base_dim = base_dim
+        self.bbox_code_size = code_size
+
+    def encode(self, locations, dimensions, orientations, img_metas):
+        """Encode CameraInstance3DBoxes by locations, dimemsions, orientations.
+
+        Args:
+            locations (Tensor): Center location for 3D boxes.
+                (N, 3)
+            dimensions (Tensor): Dimensions for 3D boxes.
+                shape (N, 3)
+            orientations (Tensor): Orientations for 3D boxes.
+                shape (N, 1)
+            img_metas (list[dict]): Meta information of each image, e.g.,
+                image size, scaling factor, etc.
+
+        Return:
+            :obj:`CameraInstance3DBoxes`
+        """
+
+        bboxes = torch.cat((locations, dimensions, orientations), dim=1)
+        assert bboxes.shape[1] == self.bbox_code_size
+        batch_bboxes = img_metas[0]['box_type_3d'](
+            bboxes, box_dim=self.bbox_code_size, origin=(0.5, 0.5, 0.5))
+
+        return batch_bboxes
+
+    def decode(self, reg, points, labels, cam_intrinsics, ratio):
+        """Decode regression into locations, dimemsions, orientations.
+
+        Args:
+            reg (Tensor): Batch regression for each predict center2d point.
+                (batch * K (max_objs), C)
+            points(Tensor): Batch projected bbox centers on image plane.
+                (batch * K (max_objs) , 2)
+            labels (Tensor): Batch predict class label for each predict
+                center2d point. (batch, K (max_objs)).
+            cam_intrinsics (Tensor): Batch images' camera intrinsic matrix.
+                (batch, 4, 4)
+            ratio (int): Scale ratio between heatmap and input image.
+
+        Return:
+            tuple(Tensor)
+        """
+        depth_offsets = reg[:, 0]
+        centers2d_offsets = reg[:, 1:3]
+        dimensions_offsets = reg[:, 3:6]
+        orientations = reg[:, 6:]
```

Comment:
A little hard code and at least should be changed to 6:7, because there can be redundant dimensions such as velocities.